lychee-lib/src/client.rs

```diff
@@ -374,38 +414,42 @@ pub struct Client {
 }
 
 impl Client {
-    /// Check a single request
+    /// Check a single request.
+    ///
+    /// `request` can be either a [`Request`] or a type that can be converted
+    /// into it. In any case, it must represent a valid URI.
     ///
     /// # Errors
     ///
-    /// This returns an `Err` if
-    /// - `request` is invalid.
-    /// - The URI of the request is invalid.
-    /// - Encrypted connection for a HTTP URL is available but unused.
-    ///   (Only checked when `Client::require_https` is `true`.)
+    /// Returns an `Err` if `request` does not represent a valid URI.
+    #[allow(clippy::missing_panics_doc)]
     pub async fn check<T, E>(&self, request: T) -> Result<Response>
     where
         Request: TryFrom<T, Error = E>,
         ErrorKind: From<E>,
     {
-        let Request { uri, source, .. } = request.try_into()?;
+        let Request {
+            ref mut uri,
+            source,
+            ..
+        } = request.try_into()?;
 
-        let uri = self.remap(uri)?;
+        self.remap(uri);
 
         // TODO: Allow filtering based on element and attribute
-        let status = if self.filter.is_excluded(&uri) {
+        let status = if self.is_excluded(uri) {
             Status::Excluded
         } else if uri.is_file() {
-            self.check_file(&uri)
+            self.check_file(uri)
         } else if uri.is_mail() {
-            self.check_mail(&uri).await
+            self.check_mail(uri).await
         } else {
-            match self.check_website(&uri).await {
+            match self.check_website(uri).await {
                 Status::Ok(code) if self.require_https && uri.scheme() == "http" => {
                     let mut https_uri = uri.clone();
-                    https_uri
-                        .set_scheme("https")
-                        .map_err(|_| ErrorKind::InvalidURI(uri.clone()))?;
+                    // here `uri` must be valid, otherwise `check_website` won't
+                    // return `Ok`, thus `set_scheme` won't fail
+                    https_uri.set_scheme("https").unwrap();
```

Comment:
I'd prefer if we kept the check. While it's true that the `uri` must be valid, this could change if we refactor the code. The HTTPS rewrite code might move into a separate function or even module and there's no guarantee that `set_scheme` will always return an `Ok` in that case.

Reply:
I think if we can add the error handling back when it's refactored. The comment here serves as the reminder.

Reply:
On the other hand, `map_err` has very little (probably zero?) overhead if it's not triggered, so why not just keep it? Then the potential error condition stays a part of the code and gets checked by the compiler, which makes refactoring less risky. If we didn't have our own `ErrorKind` for it, I'd probably opt for an `expect` with an explanation, but in our case the error variant is already there and was used before, so why not just keep it? There are quite a few error conditions in the [`set_scheme` code](https://docs.rs/url/latest/src/url/lib.rs.html#2242-2291) that are specific to the `scheme` as well and I don't think we covered that with the initial `uri` parsing, so there are cases where this code could fail where the normal parsing would succeed.

Reply:
> There are quite a few error conditions in the set_scheme code that are specific to the scheme as well and I don't think we covered that with the initial uri parsing, so there are cases where this code could fail where the normal parsing would succeed.

I've checked all cases and none of that will happen. it's not possible to fail at this point.